Add Commerce Layer Provider

BolajiAyodeji opened the pull request:
This PR adds Commerce Layer to the list of supported providers with the required hooks. This is a work in progress and is almost ready for production. Here's a brief summary of what we have currently and WIP features:

- [x] API Endpoint - Checkout
- [x] API Operations - Get all pages
- [x] API Operations - Get all products path
- [x] API Operations - Get all products
- [x] API Operations - Get product
- [x] API Operations - Get customer wishlist
- [x] API Operations - Get page
- [x] API Operations - Get product
- [x] Feature - Cart
- [x] Feature - Search
- [x] Feature - Wishlist
- [x] Feature - Customer Auth
- [x] Feature - External Checkout
- [x]  Feature - Categories
- [x] Feature - CMS integration
- [x] Provider documentation

We used the local provider as the base, and we're working on completing the pending issues and features. There will be a lot of conflicts which we will fix since we forked this some months ago. So I'll be creating this PR in draft mode, do take a look and leave comments so we know if we're in the right direction; cheers!

goncy commented on .env.template:
```diff
@@ -20,3 +20,7 @@ NEXT_PUBLIC_SWELL_PUBLIC_KEY=
 NEXT_PUBLIC_SALEOR_API_URL=
 NEXT_PUBLIC_SALEOR_CHANNEL=
+
+COMMERCELAYER_CLIENT_ID=
```
Variables here are not prefixed with NEXT_PUBLIC but the ones being used are, if those variables are being consumed in client side, they should be prefixed

BolajiAyodeji replied:
Got it! I'll fix this.

goncy commented on components/auth/LoginView.tsx:
```diff
@@ -1,6 +1,6 @@
 import { FC, useEffect, useState, useCallback } from 'react'
 import { Logo, Button, Input } from '@components/ui'
-import useLogin from '@framework/auth/use-login'
+import useLogin from '@commerce/auth/use-login'
```
No code should be added outside the provider, if something is missing on the core we can discuss it in an external PR. Also `@framework` reffers to the hook from the provider, where you can extend the one from `@commerce`

BolajiAyodeji replied:
Okay, we'll submit a separate PR for the changes outside the provider, discuss the changes, and remove the changes from this PR.

goncy reviewed (changes requested):
We can first focus on remove the code from the core and we can iterate over the rest later :)

goncy commented:
Hey @BolajiAyodeji and @acasazza thanks for updating the provider! Can you please merge with main to resolve the conflicts? Thanks!

BolajiAyodeji commented:
> Hey @BolajiAyodeji and @acasazza thanks for updating the provider! Can you please merge with main to resolve the conflicts? Thanks!

Sure, I'll do that shortly.

BolajiAyodeji commented:
You can check now @goncy, thanks.

goncy commented on framework/commercelayer/api/endpoints/checkout/checkout.ts:
```diff
@@ -0,0 +1,36 @@
+import type { CheckoutEndpoint } from '.'
+import getCredentials from '@framework/api/utils/getCredentials'
```
You shouldn't import from framework inside the provider, just relative, otherwise it may break other providers

BolajiAyodeji replied:
Alright, I'll fix that.

goncy reviewed (changes requested):
Change import from '@framework' to relative imports

BolajiAyodeji commented:
> Change import from '@framework' to relative imports

I'm not sure what file that is. Where did you encounter that error, please?

goncy commented:
> > Change import from '@framework' to relative imports
> 
> I'm not sure what file that is. Where did you encounter that error, please?

No `@framework` imports should be present inside `commercelayer` folder.

BolajiAyodeji commented:
> > > Change import from '@framework' to relative imports
> > 
> > 
> > I'm not sure what file that is. Where did you encounter that error, please?
> 
> No `@framework` imports should be present inside `commercelayer` folder.

Okay, I have fixed that.

goncy commented on framework/commercelayer/index.tsx:
```diff
@@ -0,0 +1,35 @@
+import { CommercelayerProvider } from './provider'
+import {
+  CommerceConfig,
+  CommerceProvider as CoreCommerceProvider,
```
CommerceProvider doesn't exist in `@commerce`, you want to import `CoreCommerceProvider` directly. Also it doesn't receive a `config` prop as passed below

goncy reviewed (changes requested):
Some required files are missing, like `checkout/use-checkout.tsx`, `customer/address/use-add-item` and `customer/use-add-item`, you can get the empty files from other provider like bigcommerce.

BolajiAyodeji commented:
> Some required files are missing, like `checkout/use-checkout.tsx`, `customer/address/use-add-item` and `customer/use-add-item`, you can get the empty files from other provider like bigcommerce.

Alright, I'll fix that.

goncy reviewed (changes requested):
![image](https://user-images.githubusercontent.com/6494462/147938466-b6dd0175-5809-4a98-8202-bc2b49479a6c.png)

Checkout doesn't work, the schema for `CheckoutEndpoint['handlers']` is not fulfilled.

If running `npm run build` in your local doesn't compile, we can't merge it and something is probably wrong.

To ensure you don't have an old version, delete and reinstall node_modules.

To keep CR on track, please ensure the features you expect from the provider are working correctly on your local before continuing with the review.

Have a good start of the year and let me know if you need help with something.

BolajiAyodeji commented:
Hi @goncy, thanks for the comment, and a happy new year to you!

We'll make the required changes and get back to you; cheers.

BolajiAyodeji commented:
Hi @goncy, everything works fine now and builds correctly. Do continue with the review; thank you.

goncy commented on components/wishlist/WishlistButton/WishlistButton.tsx:
```diff
@@ -31,10 +31,8 @@ const WishlistButton: FC<Props> = ({
     // @ts-ignore Wishlist is not always enabled
     (item) =>
-      item.product_id === Number(productId) &&
-      (item.variant_id as any) === Number(variant.id)
```
This change breaks all other providers types

acasazza replied:
Hi @goncy! Sorry, I don't understand that code. Our `productId` is a string, and  I can only get NaN from the Number function, and I'm not able to activate the button. How could I avoid that?

goncy replied:
In that case can you create a PR for this change separately so we can leave this PR separated from the core? And I can deploy it with other providers to see if it works

acasazza replied:
Makes sense, I will do it

Fasunle replied:
`Number(productId)` will return NaN 

You could remove the `Number(productId)` and change it to `productId`

goncy commented on components/wishlist/WishlistCard/WishlistCard.tsx:
```diff
@@ -53,6 +53,7 @@ const WishlistCard: FC<Props> = ({ item }) => {
       await addItem({
         productId: String(product.id),
         variantId: String(product.variants[0].id),
+        sizeId: String(product.variants[0].options[0].id),
```
No changes should be made to the core

goncy commented on framework/commerce/api/endpoints/login.ts:
```diff
@@ -17,7 +17,6 @@ const loginEndpoint: GetAPISchema<
   ) {
     return
   }
-
```
Remove this change to avoid changing history from the core

goncy commented on components/auth/LoginView.tsx:
```diff
@@ -20,7 +20,6 @@ const LoginView: FC<Props> = () => {
 
   const handleLogin = async (e: React.SyntheticEvent<EventTarget>) => {
     e.preventDefault()
-
```
Remove this change so the core history doesn't get affected

goncy reviewed (changes requested):
Added some comments @BolajiAyodeji @acasazza

goncy reviewed (changes requested):
The PR for the change on the core has been merged, please ensure no changes on the core has been made, including adding spaces so we can continue with the CR, thanks!

BolajiAyodeji commented:
> The PR for the change on the core has been merged, please ensure no changes on the core has been made, including adding spaces so we can continue with the CR, thanks @acasazza @BolajiAyodeji!

Hi @goncy, you can proceed with the CR. Thank you.

goncy commented on components/wishlist/WishlistButton/WishlistButton.tsx:
```diff
@@ -26,15 +26,15 @@ const WishlistButton: FC<Props> = ({
   const { data: customer } = useCustomer()
   const { openModal, setModalView } = useUI()
   const [loading, setLoading] = useState(false)
-
```
Please remove all changes, including spaces, to leave history untouched

BolajiAyodeji replied:
It's a bit hard to see changes like spaces made outside the core since we merged with `main`. Maybe because I didn't make those changes from my end. @acasazza can you take a look from yours?

acasazza commented:
Hi @goncy, I rebased our branch from the main. That should resolve all issues. Let me know. 🙌🏻

goncy commented on components/wishlist/WishlistButton/WishlistButton.tsx:
```diff
@@ -34,7 +34,7 @@ const WishlistButton: FC<Props> = ({
       item.variant_id === variant.id
   )
-
+  
```
```suggestion
 
```

BolajiAyodeji replied:
Are you suggesting that we add a space on line 37 here?

goncy commented on .env.template:
```diff
@@ -21,6 +21,9 @@ NEXT_PUBLIC_SWELL_PUBLIC_KEY=
 
+NEXT_PUBLIC_COMMERCELAYER_CLIENT_ID=
+NEXT_PUBLIC_COMMERCELAYER_ENDPOINT=
+NEXT_PUBLIC_COMMERCELAYER_MARKET_SCOPE=
```
```suggestion
NEXT_PUBLIC_COMMERCELAYER_MARKET_SCOPE=

```

goncy reviewed (changes requested):
@acasazza @bolajiayodeji I added some suggestions, also, images shouldn't live in the public folder, only the ones for the `local` provider. Please, ensure that those are being loaded from an url and add the domain to your provider next.config.js, I think we are almost there.

goncy commented on framework/commercelayer/next.config.js:
```diff
@@ -0,0 +1,8 @@
+module.exports = {
+  commerce,
+  images: {
+    domains: ['localhost', 'data.commercelayer.app'],
```
```suggestion
    domains: ['data.commercelayer.app'],
```
localhost is not needed for domains

goncy reviewed (changes requested):
Added a comment, please remove the images from public as they are not being used anymore

BolajiAyodeji commented:
Oh, yes! I forgot that, sorry.

goncy commented on framework/commercelayer/api/utils/normalizeLineItems.ts:
```diff
@@ -0,0 +1,30 @@
+          sku: attributes.sku_code,
+          price: attributes.unit_amount_float,
+          image: {
+            url: `/commercelayer_assets/${attributes.reference}_FLAT.png`,
```
This has to be updated to use the external image

goncy reviewed (changes requested):
Update references to public images

goncy commented:
@BolajiAyodeji 
![image](https://user-images.githubusercontent.com/6494462/151970408-990155b8-e3d7-4843-8df7-9adc1219de73.png)

When clicking an element on the cart it redirects to an undefined route

goncy commented on framework/commercelayer/api/utils/normalizeLineItems.ts:
```diff
@@ -1,3 +1,5 @@
+import data from '../../data.json'
```
The schema should be built with items from the real services, otherwise it won't work when people deploy their apps.

goncy commented on framework/commercelayer/api/operations/get-product.ts:
```diff
@@ -1,8 +1,8 @@
 import { GetProductOperation } from '@commerce/types/product'
-import data from '../../data.json'
 import type { OperationContext } from '@commerce/api/operations'
+import data from '../../data.json'
```
Products should come from the api.

goncy reviewed (changes requested):
Please, ensure that all changes implemented on the provider works when people deploy their own version with their own store. No hardcoded values for the demo should be on providers code.

BolajiAyodeji commented:
> Please, ensure that all changes implemented on the provider works when people deploy their own version with their own store. No hardcoded values for the demo should be on providers code.

Hi @goncy,

Thanks for the review and feedback!

So Commerce Layer [doesn't manage content](https://commercelayer.io/docs/core-concepts/content-vs-commerce). As a fully headless platform, we only handle transactional functionalities while content is managed independently by context experts on a content management system. This allows brands to utilize any content model, produce a better customer experience, outstand competition, and unleash creativity.

Since the provider doesn't support CMS's yet, we had to resort to using the local content method for now pending till when we can integrate our provider framework with available CMS. Our [documentation](https://github.com/BolajiAyodeji/commerce/tree/commercelayer-provider/framework/commercelayer#readme) will cover this and all new users have to do is either use the provider as a boilerplate and integrate their own CMS or rely on the local content method. Basically, the content comes from the local file but the line item comes from the API and both are connected via the SKU.

Do let us know how to proceed, please. Thanks.

BolajiAyodeji commented:
Hi @goncy and @okbel, this PR is ready for another review now. Basically we have added the content as an external resource as discussed and updated our package and documentation accordingly. Please take a look and let us know the next steps. Thank you and have a great weekend!

leerob commented:
Hey there! Thank you for your contribution. We have decided to take Next.js Commerce in a new direction and will be closing out current PRs and issues due to this change. Please see this PR for more details: https://github.com/vercel/commerce/pull/966

Would love to see a Commerce LAyer version of v2!
